proposals/gc/MVP.md

```diff
@@ -0,0 +1,272 @@
+  - `module ::= {..., types vec(<deftype>)}`
+
+* `structtype` describes a structure with statically indexed fields
+  - `structtype ::= struct <fieldtype>*`
+
+* `arraytype` describes an array with dynamically indexed fields
+  - `arraytype ::= array <fieldtype>`
+
+* `fieldtype` describes a struct or array field and whether it is mutable
+  - `fieldtype ::= <mutability> <storagetype>`
+  - `storagetype ::= <valtype> | <packedtype>`
+  - `packedtype ::= i8 | i16`
+
+* Unpacking a storage type yields `i32` for packed types, otherwise the type itself
+  - `unpacked(t) = t`
+  - `unpacked(pt) = i32`
+
+
+#### Imports
+
+* `type <typetype>` is an import description with an upper bound
+  - `importdesc ::= ... | type <reftype>`
+  - Note: `type` may get additional parameters in the future
+
+
+#### Exports
+
+* `type <typeidx>` is an export description
+  - `exportdesc ::= ... | type <typeidx>`
+  - `type $t ok` iff `$t` is defined in the context
+
+
+#### Subtyping
+
+Greatest fixpoint (co-inductive interpretation) of the given rules (implying reflexivity and transitivity).
+
+* `eqref` is a subtype of `anyref`
+  - `eqref <: anyref`
+  - Note: `int31ref` and `anyfunc` are *not* a subtypes of `eqref`, i.e., those types do not expose reference equality
+
+* `nullref` is a subtype of `eqref`
+  - `nullref <: eqref`
+  - Note: `nullref` is *not* a subtype of `anyfunc`, `int31ref`or any concrete reference type, i.e., those types are not nullable
+
+* `int31ref` is a subtype of `anyref`
+  - `int31ref <: anyref`
+
+* Any concrete reference type is a subtype of `anyref`
+  - `ref $t <: anyref`
+  - Note: concrete reference types are *not* a supertypes of `nullref`, i.e., nut nullable
+
+* Any function reference type is a subtype of `anyfunc`
+  - `ref $t <: anyfunc`
+     - iff `$t = <functype>`
+
+* Any concrete reference type is a subtype of `eqref` if its not a function
+  - `ref $t <: eqref`
+     - if `$t = <structtype>` or `$t = <arraytype>`
+     - or `$t = type rt` and `rt <: eqref`
+  - TODO: provide a way to make data types non-eq, especially immutable ones
+
+* Concrete reference types are covariant
+  - `ref $t1 <: ref $t2`
+     - iff `$t1 <: $t2`
+
+* Structure types support width and depth subtyping
+  - `struct <fieldtype1>* <fieldtype1'>* <: struct <fieldtype2>*`
+    - iff `(<fieldtype1> <: <fieldtype2>)*`
+
+* Array types support depth subtyping
+  - `array <fieldtype1> <: array <fieldtype2>`
+    - iff `<fieldtype1> <: <fieldtype2>`
+
+* Field types are covariant if they are immutable, invariant otherwise
+  - `const <valtype1> <: const <valtype2>`
+    - iff `<valtype1> <: <valtype2>`
+  - `var <valtype> <: var <valtype>`
+  - Note: mutable fields are *not* subtypes of immutable ones, so `const` really means constant, not read-only
+
+
+#### Defaultability
+
+* Any numeric value type is defaultable
+
+* A reference value type is defaultable if it is not of the form `ref $t`
+
+* Locals must have a type that is defaultable.
```

Comment:
Without adding a `nullable` type constructor, it seems like any source language with ubiquitous nullable types will be forced to use tons of otherwise-unnecessary `anyref`s and `cast`s.  Nullability seems like an important thing not to bolt on later too, so can we add it?

Reply:
I think it could be added, but there is a risk though that people will be lazy and default to using `optref` instead of `anyref` since it gives "easier" interop between all sorts of languages, which will make languages that are null-safe pay the price for having nulls. A world where non-null is default and languages that are liberal with null pay the price makes more sense to me, and will promote more robust software ecosystems built on wasm.

Reply:
I am on the fence here. The main benefit of a nullable ref type over anyref is that a downcast to the non-nullable concrete ref type is gonna be slightly cheaper. But you'd still need it. (Or are you suggesting that all ref instructions should also work with nullable refs directly?)

Reply:
A downcast from a `nullref $t` to `ref $t` is going to be significantly cheaper than a downcast from `anyref`; the former is a single null check; the latter several instructions and memory access (several, if TLS is used to avoid baking in pointers to the code).

Reply:
Okay, added `optref` (for a short name). (I abandoned the idea for a separate `nullable` type constructor because it actually introduces more anomalies than benefits.)

Reply:
Thanks, looks good!

Reply:
Oh: perhaps also mention table element kinds (or maybe this is solved by validation constraint on `table.grow` and adding a new `table.grow_init` that takes a value to initialize with?

Reply:
I think `table.grow` should take an initialisation operand right away, no point in having two separate instructions there. (I just realised that this is instruction is not currently part of any proposal, so I opened an issue for the reference types proposal.)

But there is another constraint we need: a table *definition* of non-zero initial size must have a defaultable element type. Added that.